File: KPCTabsControlDemo/PaneViewController.m
//
//  PaneViewController.m
//  KPCTabsControl
//
//  Created by Cédric Foellmi on 31/10/14.
//  Copyright (c) 2014 Cédric Foellmi. All rights reserved.
//

#import "PaneViewController.h"
#import "KPCTabsControlConstants.h"

@interface PaneViewController ()
@property (strong) NSArray *titles;
@property (strong) NSDictionary *icons;
@property (strong) NSDictionary *menus;
@end

@implementation PaneViewController

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self setup];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        [self setup];
    }
    return self;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        [self setup];
    }
    return self;
}

- (void)setup
{
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(updateLabelsUponReframe:)
                                                 name:NSViewFrameDidChangeNotification
                                               object:self.tabsBar];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(updateUponPaneSelectionDidChange:)
                                                 name:@"PaneSelectionDidChangeNotification"
                                               object:nil];
    
    switch ([self.title isEqualToString:@"pane1"]) {
        case 0: {
            self.titles = @[@"Tab 1", @"Tab 2", @"Tab 3", @"Tab 4", @"Tab 5"];
            NSMenu *tab2Menu = [[NSMenu alloc] init];
            [tab2Menu addItemWithTitle:@"Action 1" action:NULL keyEquivalent:@""];
            [tab2Menu addItemWithTitle:@"Action 2" action:NULL keyEquivalent:@""];
            self.menus = @{@"Tab 2": tab2Menu};
            break;
        }

        case 1:
            self.titles = @[@"Tab a", @"Tab b", @"Tab c", @"Tab d"];
            break;

        default:
            break;
    }
}

- (void)awakeFromNib
{
    [super awakeFromNib];

	if ([self.title isEqualToString:@"pane2"]) {
		self.tabsBar.maxTabWidth = 130.0;
		self.tabsBar.minTabWidth = 100.0;
	}

	NSString *labelString = [NSString stringWithFormat:@"min %.0f < %.0f < %.0f max",
							 self.tabsBar.minTabWidth, self.tabsBar.currentTabWidth, self.tabsBar.maxTabWidth];

	[self.tabWidthsLabel setStringValue:labelString];

	[self.tabsBar setPreferFullWidthTabs:self.useFullWidthTabsCheckButton.state];
    [self.tabsBar reloadTabs];
}

- (IBAction)toggleFullWidthTabs:(id)sender
{
	[self.tabsBar setPreferFullWidthTabs:self.useFullWidthTabsCheckButton.state];
}

- (void)mouseDown:(NSEvent *)theEvent
{
    [super mouseDown:theEvent];
    
    BOOL sendNotification = (!self.tabsBar.isHighlighted);

    [self.tabsBar highlight:YES];
    
    if (sendNotification) {
        [[NSNotificationCenter defaultCenter] postNotificationName:@"PaneSelectionDidChangeNotification"
                                                            object:self
                                                          userInfo:nil];
    }
}

- (void)updateUponPaneSelectionDidChange:(NSNotification *)notif
{
    if ([notif object] != self) {
        [self.tabsBar highlight:NO];
    }
}

- (void)updateLabelsUponReframe:(NSNotification *)notif
{
    NSString *labelString = [NSString stringWithFormat:@"min %.0f < %.0f < %.0f max",
                             self.tabsBar.minTabWidth, self.tabsBar.currentTabWidth, self.tabsBar.maxTabWidth];
    
    [self.tabWidthsLabel setStringValue:labelString];
}

#pragma mark - KPCTabsControlDataSource

- (NSUInteger)tabsControlNumberOfTabs:(KPCTabsControl *)tabControl
{
    return self.titles.count;
}

- (id)tabsControl:(KPCTabsControl *)tabControl itemAtIndex:(NSUInteger)index
{
    return self.titles[index];
}

- (NSString *)tabsControl:(KPCTabsControl *)tabControl titleForItem:(id)item
{
    NSUInteger index = [self.titles indexOfObject:item];
    return (index == NSNotFound) ? @"?" : self.titles[index];
}


#pragma - mark Optionals

- (NSMenu *)tabsControl:(KPCTabsControl *)tabControl menuForItem:(id)item
{
    return [self.menus objectForKey:item];
}

- (NSImage *)tabsControl:(KPCTabsControl *)tabControl iconForItem:(id)item
{
    if ([item isEqualToString:@"Tab a"]) {
        return [NSImage imageNamed:@"Star"];
    }
    else if ([item isEqualToString:@"Tab b"]) {
        return [NSImage imageNamed:@"Oval"];
    }
    else if ([item isEqualToString:@"Tab 2"]) {
        return [NSImage imageNamed:@"Triangle"];
    }
    else if ([item isEqualToString:@"Tab 3"]) {
        return [NSImage imageNamed:@"Spiral"];
    }
    else if ([item isEqualToString:@"Tab 4"]) {
        return [NSImage imageNamed:@"Polygon"];
    }
    
    return nil;
}

#pragma - mark KPCTabsControlDelegate

- (BOOL)tabsControl:(KPCTabsControl *)tabControl canReorderItem:(id)item
{
	return YES;
}

- (void)tabsControl:(KPCTabsControl *)tabControl didReorderItems:(NSArray *)itemArray
{
    self.titles = itemArray;
}

- (BOOL)tabsControl:(KPCTabsControl *)tabControl canEditTitleOfItem:(id)item
{
    return YES;
}

- (void)tabsControl:(KPCTabsControl *)tabControl setTitle:(NSString *)title forItem:(id)item
{
    NSLog(@"-> %@", title);
    NSUInteger index = [self.titles indexOfObject:item];
    NSMutableArray *newTitles = [self.titles mutableCopy];
    [newTitles replaceObjectAtIndex:index withObject:title];
    self.titles = [newTitles copy];
}


@end

File: KPCTabsControlDemo/PaneViewController.swift
//
//  ViewController.swift
//  KPCTabsControlDemo
//
//  Created by Cédric Foellmi on 15/07/16.
//  Licensed under the MIT License (see LICENSE file)
//

import Cocoa
import KPCTabsControl

// We need a class (rather than a struct or a tuple – which would be nice) because TabsControlDelegate has
// @optional methods. To have such optionaling, we need to mark the protocol as @objc. With such marking,
// we can't have pure-Swift 'Any' return object or argument. Buh...

class Item {
    var title: String = ""
    var icon: NSImage?
    var menu: NSMenu?
    var altIcon: NSImage?
    var selectable: Bool
    
    init(title: String, icon: NSImage?, menu: NSMenu?, altIcon: NSImage?, selectable: Bool = true) {
        self.title = title
        self.icon = icon
        self.menu = menu
        self.altIcon = altIcon
        self.selectable = selectable
    }
}

extension Item: Equatable { }

func ==(lhs: Item, rhs: Item) -> Bool {
    return lhs.title == rhs.title
}

class PaneViewController: NSViewController, TabsControlDataSource, TabsControlDelegate {

    @IBOutlet weak var tabsBar: TabsControl?
    @IBOutlet weak var useFullWidthTabsCheckButton: NSButton?
    @IBOutlet weak var tabWidthsLabel: NSTextField?

    var items: Array<Item> = []
    override var title: String? {
        didSet { self.tabWidthsLabel?.stringValue = self.title! }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.tabsBar?.dataSource = self
        self.tabsBar?.delegate = self
        self.tabsBar?.reloadTabs()
    }
        
    // MARK: TabsControlDataSource
    
    func tabsControlNumberOfTabs(_ control: TabsControl) -> Int {
        return self.items.count
    }
    
    func tabsControl(_ control: TabsControl, itemAtIndex index: Int) -> AnyObject {
        return self.items[index]
    }
    
    func tabsControl(_ control: TabsControl, titleForItem item: AnyObject) -> String {
        return (item as! Item).title
    }
    
    // MARK: TabsControlDataSource : Optionals
    
    func tabsControl(_ control: TabsControl, menuForItem item: AnyObject) -> NSMenu? {
        return (item as! Item).menu
    }
    
    func tabsControl(_ control: TabsControl, iconForItem item: AnyObject) -> NSImage? {
        return (item as! Item).icon
    }
    
    func tabsControl(_ control: TabsControl, titleAlternativeIconForItem item: AnyObject) -> NSImage? {
        return (item as! Item).altIcon
    }

    // MARK: TabsControlDelegate
    
    func tabsControl(_ control: TabsControl, canReorderItem item: AnyObject) -> Bool {
        return true
    }
    
    func tabsControl(_ control: TabsControl, didReorderItems items: [AnyObject]) {
        self.items = items.map { $0 as! Item }
    }
    
    func tabsControl(_ control: TabsControl, canEditTitleOfItem: AnyObject) -> Bool {
        return true
    }
    
    func tabsControl(_ control: TabsControl, setTitle newTitle: String, forItem item: AnyObject) {
        let typedItem = item as! Item
        let titles = self.items.map { $0.title }
        let index = titles.index(of: typedItem.title)!

        let newItem = Item(title: newTitle, icon: typedItem.icon, menu: typedItem.menu, altIcon: typedItem.altIcon)
        let range = index..<index+1
        self.items.replaceSubrange(range, with: [newItem])
    }

    func tabsControl(_ control: TabsControl, canSelectItem item: AnyObject) -> Bool {
        return (item as! Item).selectable
    }
}

